### PR TITLE
Added support for enable/disable of DML triggers

### PR DESCRIPTION
**Pull request by rohit01010**
### Description
Currently we do not support enabling/disabling of DML triggers. This PR will add support for enabling/disabling of the DML triggers on table with the following syntax:

```
ENABLE TRIGGER { [ schema_name . ] trigger_name [ ,...n ] | ALL }  
ON { table_name }  [ ; ]  
```
```
DISABLE TRIGGER { [ schema_name . ] trigger_name [ ,...n ] | ALL }  
ON { table_name } [ ; ]  
```
```
ALTER TABLE { database_name.schema_name.table_name | schema_name.table_name | table_name }
{ ENABLE | DISABLE } TRIGGER { ALL | trigger_name [ ,...n ] }
```

Signed-off-by: Rohit Bhagat [rohitbgt@amazon.com](mailto:rohitbgt@amazon.com)

### Issues Resolved
BABEL-3326

### Test Scenarios Covered ###
* **Use case based -** YES

* **Boundary conditions -** NA

* **Arbitrary inputs -** NA

* **Negative test cases -** NA

* **Minor version upgrade tests -** NA

* **Major version upgrade tests -** NA

* **Performance tests -** NA

* **Tooling impact -** NA

* **Client tests -** NA

### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: test/JDBC/expected/BABEL-3326-vu-prepare.out [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+),
+(   'Temp Name4',
+    'Temp Address 4',
+    10000
+)
+GO
+~~ROW COUNT: 4~~
+
+
+CREATE TRIGGER [dbo].[TR_ins_babel_3326_Employees] ON [dbo].[babel_3326_Employees]
```
Comment (Deepesh125) [inline test/JDBC/expected/BABEL-3326-vu-prepare.out]: use multibyte chars in trigger name

File: test/JDBC/expected/BABEL-3326-vu-verify.out [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+GO
+~~START~~
+varchar
+Entry Inserted
+~~END~~
+
+~~ROW COUNT: 1~~
+
+
+ALTER TABLE babel_3326_Employees DISABLE TRIGGER TR_ins_babel_3326_Employees
```
Comment (Deepesh125) [inline test/JDBC/expected/BABEL-3326-vu-verify.out]: what if you write query like, `ALTER TABLE "DROP TRIGGER TR_ins_babel_3326_Employees; babel_3326_Employees" DISABLE TRIGGER TR_ins_babel_3326_Employees`

File: test/JDBC/expected/BABEL-3326-vu-verify.out [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
+GO
+
+-- TR_ins_babel_3326_Employees should be disabled
+INSERT INTO dbo.babel_3326_Employees ( EmployeeName , EmployeeAddress ,MonthSalary )
+VALUES ( 'Temp Name5' , 'Address Name 5', 4000)
+GO
+~~ROW COUNT: 1~~
+
+
+ALTER TABLE babel_3326_Employees ENABLE TRIGGER TR_ins_babel_3326_Employees
```
Comment (Deepesh125) [inline test/JDBC/expected/BABEL-3326-vu-verify.out]: use multi trigger case is missing for ENABLE, DISABLE and ALTER TABLE case

**Review by Deepesh125 — CHANGES_REQUESTED** [review]

**Review by robverschoor — CHANGES_REQUESTED** [review]
Please add more test cases:
- ALTER TABLE can operate cross-database
- use mixed-case identifiers, with different case when created  than when enabled/disabled
- use bracketed identifiers, with different case when created  than when enabled/disabled
- test cases for not-supported variants (if not already present somewhere else)
- test case where non-owner user of the table has select/insert/update/delete access but should not be allowed to enable/disable a trigger
- drop a disabled trigger
- use

File: contrib/babelfishpg_tsql/src/tsqlIface.cpp [hunk 4]
```diff
@@ -1459,6 +1459,159 @@ class tsqlBuilder : public tsqlCommonMutator
...
+			 * When we come across a alter table query which enable/disable trigger with multiple trigger name, 
+			 * we will replace it with list of alter table statements, a statement for each trigger.
+			 * As Postgres only support enabling/disabling of only one trigger using alter table syntax, so a call like:
+			 * 
+			 * 	ALTER TABLE Employees { ENABLE | DISABLE } TRIGGER trigger_a, trigger_b
+			 *	GO
+			 * 
+			 * will be re-written as:
+			 * 
+			 * 	ALTER TABLE Employees { ENABLE | DISABLE } TRIGGER trigger_a; ALTER TABLE Employees { ENABLE | DISABLE } TRIGGER trigger_b;
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/tsqlIface.cpp]: how do we handle failures here? what if for alter1; alter2; alter3 alter1 fails and alter2 fails.

File: test/JDBC/expected/BABEL-3326-vu-prepare.out (hunk 1 above)
Comment (rohit01010) [reply test/JDBC/expected/BABEL-3326-vu-prepare.out]: Resolved

File: test/JDBC/expected/BABEL-3326-vu-verify.out (hunk 3 above)
Comment (rohit01010) [reply test/JDBC/expected/BABEL-3326-vu-verify.out]: Resolved

File: contrib/babelfishpg_tsql/src/tsqlIface.cpp (hunk 4 above)
Comment (rohit01010) [reply contrib/babelfishpg_tsql/src/tsqlIface.cpp]: All these statements, alter1; alter2; alter3 are given as batch of statements, when any one of them fails, the whole query will fail.

File: test/JDBC/expected/BABEL-3326-vu-verify.out (hunk 2 above)
Comment (rohit01010) [reply test/JDBC/expected/BABEL-3326-vu-verify.out]: It will throw an error as follows 
\~\~ERROR (Message: relation "drop trigger tr_ins_babel_3326_employees; babel_3326_employees" does not exist)\~\~

File: contrib/babelfishpg_tsql/src/pl_handler.c [hunk 5]
```diff
@@ -590,6 +590,53 @@ pltsql_pre_parse_analyze(ParseState *pstate, RawStmt *parseTree)
...
+					rel_schema = get_logical_schema_name(atstmt->relation->schemaname, true);
+
+					if (rel_schema == NULL)
+					{
+						rel_schema = atstmt->relation->schemaname;
+					}
+				}
+				else
+				{
+					rel_schema = "dbo";
```
Comment (sumitj824) [inline contrib/babelfishpg_tsql/src/pl_handler.c]: User can change the default schema, instead of hard-coding it to `dbo` we should lookup for the default schema of the current user. Please check with T-SQL and update here accordingly.

File: contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-rule.y [hunk 6]
```diff
@@ -475,6 +475,72 @@ tsql_AlterLoginStmt:
...
+						}
+						else
+						{
+							n1->subtype = AT_DisableTrig;
+						}
+						
+						if (list_length(lst) > 1)
+						{
+							n1->name = strcat(((String *)list_nth(lst,0))->sval, ".");
+							n1->name = strcat(n1->name, ((String *)list_nth(lst,1))->sval);
```
Comment (sumitj824) [inline contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-rule.y]: I think we can use `psprintf` here for these two directly, `strcat` is not safe to use.

**Review by sumitj824 — CHANGES_REQUESTED** [review]

File: test/JDBC/input/pg_stat_statements_tsql.mix [hunk 7]
```diff
@@ -158,7 +158,7 @@ go
 -- enable/disable trigger 
 ALTER TABLE pgss_test ENABLE TRIGGER pgss_trigger1
 go
-ALTER TABLE pgss_test DISABLE  TRIGGER pgss_trigger1
+ALTER TABLE pgss_test DISABLE TRIGGER pgss_trigger1
```
Comment (sumitj824) [inline test/JDBC/input/pg_stat_statements_tsql.mix]: why this change?

File: contrib/babelfishpg_tsql/src/pl_handler.c (hunk 5 above)
Comment (rohit01010) [reply contrib/babelfishpg_tsql/src/pl_handler.c]: Updated

File: contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-rule.y (hunk 6 above)
Comment (rohit01010) [reply contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-rule.y]: replaced `strcat` with `psprintf`

File: test/JDBC/input/pg_stat_statements_tsql.mix (hunk 7 above)
Comment (rohit01010) [reply test/JDBC/input/pg_stat_statements_tsql.mix]: Removed

File: test/JDBC/input/BABEL-3326-vu-verify.mix [hunk 8]
```diff
@@ -0,0 +1,527 @@
...
+INSERT INTO dbo.babel_3326_Employees ( EmployeeName, EmployeeAddress, MonthSalary )
+VALUES ( 'Temp Name5' , 'Address Name 5', 4000)
+GO
+
+UPDATE dbo.babel_3326_Employees
+SET MonthSalary = 3333
+WHERE EmployeeID = 2;
+GO
+
+ENABLE TRIGGER [TR_ins_안녕하세요_babel_3326_Employees], TR_upd_abcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyz_babel_3326_Employees ON babel_3326_Employees
```
Comment (forestkeeper) [inline test/JDBC/input/BABEL-3326-vu-verify.mix:200]: For MS SQL, what's the result for : 

begin transaction
disable trigger
rollback 

Just wanted to make sure this DDL can be part of transaction

Comment (forestkeeper) [reply test/JDBC/input/BABEL-3326-vu-verify.mix:200]: Will that disable rollback by transaction in this case ?

Comment (rohit01010) [reply test/JDBC/input/BABEL-3326-vu-verify.mix:200]: In MS SQL,  the result for:

```
begin transaction
disable trigger
rollback
```
is: the effect of disable trigger will also be rollbacked.

**Review by forestkeeper — DISMISSED** [review]
LGTM

File: contrib/babelfishpg_tsql/src/pl_handler.c [hunk 9]
```diff
@@ -590,6 +590,53 @@ pltsql_pre_parse_analyze(ParseState *pstate, RawStmt *parseTree)
...
+		const char *rel_schema;
+
+		foreach(lc, atstmt->cmds)
+		{
+			AlterTableCmd *cmd = (AlterTableCmd *)lfirst(lc);
+			if (cmd->subtype == AT_EnableTrig || cmd->subtype == AT_DisableTrig)
+			{
+				if (strchr(cmd->name, '.') != NULL)
+				{
+					trig_schema = strtok(cmd->name, ".");
```
Comment (Deepesh125) [inline contrib/babelfishpg_tsql/src/pl_handler.c]: 1. strtok may modify input string, cmd->name so we should use strtok on duplicated string to avoid rewriting original cmd->name.  
2. what would happen when you have trigger name something like "sch.ema"."tri.gg.er"

File: contrib/babelfishpg_tsql/src/pl_handler.c [hunk 10]
```diff
@@ -590,6 +590,53 @@ pltsql_pre_parse_analyze(ParseState *pstate, RawStmt *parseTree)
...
+				else
+				{
+					trig_schema = NULL;
+				}
+				
+				if (atstmt->relation->schemaname != NULL)
+				{
+					rel_schema = get_logical_schema_name(atstmt->relation->schemaname, true);
+
+					if (rel_schema == NULL)
```
Comment (Deepesh125) [inline contrib/babelfishpg_tsql/src/pl_handler.c]: in which case it could be NULL?

File: contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-rule.y [hunk 11]
```diff
@@ -475,6 +475,71 @@ tsql_AlterLoginStmt:
...
+							n1->subtype = AT_EnableTrig;
+						}
+						else
+						{
+							n1->subtype = AT_DisableTrig;
+						}
+						
+						if (list_length(lst) > 1)
+						{
+							n1->name = psprintf("%s.%s", ((String *)list_nth(lst,0))->sval, ((String *)list_nth(lst,1))->sval);
```
Comment (Deepesh125) [inline contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-rule.y]: In case of "sch.ema"."tri.gg.er" how would n1->name look?

File: contrib/babelfishpg_tsql/src/pl_handler.c [hunk 12]
```diff
@@ -590,6 +590,53 @@ pltsql_pre_parse_analyze(ParseState *pstate, RawStmt *parseTree)
...
+					{
+						rel_schema = atstmt->relation->schemaname;
+					}
+				}
+				else
+				{
+					rel_schema = get_authid_user_ext_schema_name(get_cur_db_name(), GetUserNameFromId(GetUserId(), false));
+				}
+
+				if (trig_schema != NULL && strcasecmp(trig_schema, rel_schema) != 0)
```
Comment (Deepesh125) [inline contrib/babelfishpg_tsql/src/pl_handler.c]: use strncmp with length = min (trig_schema, rel_schema)

File: contrib/babelfishpg_tsql/src/pl_handler.c [hunk 13]
```diff
@@ -590,6 +590,53 @@ pltsql_pre_parse_analyze(ParseState *pstate, RawStmt *parseTree)
...
+	{
+		AlterTableStmt *atstmt = (AlterTableStmt *) parseTree->stmt;
+		ListCell *lc;
+		char *trig_schema;
+		const char *rel_schema;
+
+		foreach(lc, atstmt->cmds)
+		{
+			AlterTableCmd *cmd = (AlterTableCmd *)lfirst(lc);
+			if (cmd->subtype == AT_EnableTrig || cmd->subtype == AT_DisableTrig)
```
Comment (Deepesh125) [inline contrib/babelfishpg_tsql/src/pl_handler.c:603]: how are we handling ENABLE\DISABLE ALL?

File: contrib/babelfishpg_tsql/src/pl_handler.c [hunk 14]
```diff
@@ -590,6 +590,53 @@ pltsql_pre_parse_analyze(ParseState *pstate, RawStmt *parseTree)
...
+					rel_schema = get_logical_schema_name(atstmt->relation->schemaname, true);
+
+					if (rel_schema == NULL)
+					{
+						rel_schema = atstmt->relation->schemaname;
+					}
+				}
+				else
+				{
+					rel_schema = get_authid_user_ext_schema_name(get_cur_db_name(), GetUserNameFromId(GetUserId(), false));
```
Comment (Deepesh125) [inline contrib/babelfishpg_tsql/src/pl_handler.c:619]: should we free rel_schema?

**Review by Deepesh125 — CHANGES_REQUESTED** [review]

File: contrib/babelfishpg_tsql/src/pl_handler.c (hunk 14 above)
Comment (rohit01010) [reply contrib/babelfishpg_tsql/src/pl_handler.c:619]: added code to free the rel_schema

File: contrib/babelfishpg_tsql/src/pl_handler.c (hunk 13 above)
Comment (rohit01010) [reply contrib/babelfishpg_tsql/src/pl_handler.c:603]: no need to add any similar handling for ENABLE\DISABLE ALL here as,
`ENABLE/DISABLE TRIGGER ALL ON <table>` will enable/disable all the triggers present on that table
where as in case of `ENABLE/DISABLE TRIGGER <schema>.<trigger> ON <schema>.<table>`, we need to check whether both table and trigger are having same schema or not

File: contrib/babelfishpg_tsql/src/pl_handler.c (hunk 12 above)
Comment (rohit01010) [reply contrib/babelfishpg_tsql/src/pl_handler.c]: done

File: contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-rule.y (hunk 11 above)
Comment (rohit01010) [reply contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-rule.y]: changed logic to store schema name instead of storing it in n1->name

File: contrib/babelfishpg_tsql/src/pl_handler.c (hunk 10 above)
Comment (rohit01010) [reply contrib/babelfishpg_tsql/src/pl_handler.c]: refactored code

File: contrib/babelfishpg_tsql/src/pl_handler.c (hunk 9 above)
Comment (rohit01010) [reply contrib/babelfishpg_tsql/src/pl_handler.c]: changed logic to store schema name instead of storing it in cmd->name with delimeter `.`

File: contrib/babelfishpg_tsql/src/pltsql.h [hunk 15]
```diff
@@ -1704,6 +1704,24 @@ typedef struct PLwdatum
...
+								 * number */
+	RoleSpec   *newowner;
+	Node	   *def;			/* definition of new column, index,
+								 * constraint, or parent table */
+	DropBehavior behavior;		/* RESTRICT or CASCADE for DROP cases */
+	bool		missing_ok;		/* skip error if missing? */
+	bool		recurse;		/* exec-time recursion */
+	
+	char	   *schemaname;
+} AlterTableCmd_tsql;
```
Comment (Deepesh125) [inline contrib/babelfishpg_tsql/src/pltsql.h]: As discussed offline, its possible to inline AlterTableCmd struct here which would help us in better maintainability in long term

Comment (rohit01010) [reply contrib/babelfishpg_tsql/src/pltsql.h]: done

**Review by forestkeeper — APPROVED** [review]

File: contrib/babelfishpg_tsql/src/pl_handler.c (hunk 9 above)
Comment (forestkeeper) [reply contrib/babelfishpg_tsql/src/pl_handler.c]: Could you resolve this conversation with Dipesh and I'll merge the change.

**Review by Deepesh125 — APPROVED** [review]

File: contrib/babelfishpg_tsql/src/pl_handler.c [hunk 16]
```diff
@@ -590,6 +590,50 @@ pltsql_pre_parse_analyze(ParseState *pstate, RawStmt *parseTree)
...
+				if (atstmt->relation->schemaname != NULL)
+				{
+					rel_schema = atstmt->relation->schemaname;
+				}
+				else
+				{
+					rel_schema = get_authid_user_ext_schema_name(get_cur_db_name(), GetUserNameFromId(GetUserId(), false));
+				}
+
+				if (trig_schema != NULL && strncmp(trig_schema, rel_schema, Min(strlen(rel_schema), strlen(trig_schema))) != 0)
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/pl_handler.c:622]: how will this work for cases where schema1 is a substr of schema2? like abc and abcd
